pkg/agent/proxy/proxier.go

```diff
@@ -157,74 +158,89 @@ func (p *proxier) removeStaleServices() {
-				klog.ErrorS(err, "Error when removing Endpoints flows for Service", "ServicePortName", svcPortName, "ServiceInfo", svcInfoStr)
-				continue
-			}
-			delete(p.endpointsInstalledMap, svcPortName)
-		}
-		// Remove NodePort and ClusterIP flows and configurations.
-		if p.proxyAll {
-			if svcInfo.NodePort() > 0 {
-				if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
-					klog.ErrorS(err, "Error when uninstalling NodePort flows and configurations for Service", "ServicePortName", svcPortName, "ServiceInfo", svcInfoStr)
-					continue
-				}
-			}
-		}
-		// Remove LoadBalancer flows and configurations.
-		if p.proxyLoadBalancerIPs && len(svcInfo.LoadBalancerIPStrings()) > 0 {
-			if err := p.uninstallLoadBalancerService(svcInfoStr, svcInfo.LoadBalancerIPStrings(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
-				klog.ErrorS(err, "Error when uninstalling LoadBalancer flows and configurations for Service", "ServicePortName", svcPortName, "ServiceInfo", svcInfoStr)
-				continue
-			}
-		}
 		// Remove Service group which has only local Endpoints.
-		if groupID, exist := p.groupCounter.Get(svcPortName, true); exist {
-			if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
-				klog.ErrorS(err, "Error when uninstalling group of local Endpoints for Service", "ServicePortName", svcPortName, "ServiceInfo", svcInfoStr)
-				continue
-			}
-			p.groupCounter.Recycle(svcPortName, true)
+		if !p.removeServiceGroup(svcPortName, true) {
+			continue
 		}
 		// Remove Service group which has all Endpoints.
-		if groupID, exist := p.groupCounter.Get(svcPortName, false); exist {
-			if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
-				klog.ErrorS(err, "Error when uninstalling group of all Endpoints for Service", "ServicePortName", svcPortName, "ServiceInfo", svcInfoStr)
+		if !p.removeServiceGroup(svcPortName, false) {
+			continue
+		}
+		// Remove associated Endpoints flows.
+		if endpoints, ok := p.endpointsInstalledMap[svcPortName]; ok {
+			if !p.removeStaleEndpoints(svcPortName, svcInfo.OFProtocol, endpoints) {
 				continue
 			}
-			p.groupCounter.Recycle(svcPortName, false)
+			delete(p.endpointsInstalledMap, svcPortName)
 		}
 
 		delete(p.serviceInstalledMap, svcPortName)
 		p.deleteServiceByIP(svcInfoStr)
 	}
 }
 
-func getBindingProtoForIPProto(isIPv6 bool, protocol corev1.Protocol) binding.Protocol {
-	var bindingProtocol binding.Protocol
-	if isIPv6 {
-		bindingProtocol = binding.ProtocolTCPv6
-		if protocol == corev1.ProtocolUDP {
-			bindingProtocol = binding.ProtocolUDPv6
-		} else if protocol == corev1.ProtocolSCTP {
-			bindingProtocol = binding.ProtocolSCTPv6
+func (p *proxier) removeServiceFlows(svcInfo *types.ServiceInfo) bool {
+	svcInfoStr := svcInfo.String()
+	if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
```

Comment:
Maybe add some comments here. This is used to ClusterIP flows.

Reply:
done